Model submission for BOSDYN_SPOT and two sensor configs of CTU-CRAS-Norlab

peci1 opened the pull request:
This PR adds Boston Dynamics Spot robot. It adds robot BOSDYN_SPOT which is the bare platform with no additional sensors. On top of this common base, two sensor configs CTU_CRAS_NORLAB_SPOT_SENSOR_CONFIG_X are submitted.

It is expected that the bare robot model will also be utilized by other teams when submitting their own sensor configs of Spot. We're coordinating the effort together in https://github.com/ctu-vras/subt/discussions/1 . It is not expected that the base robot would be used alone without any sensor config - that's why it doesn't have the _sensor_config_x suffix. Sensor config 1 and 2 differ just in the (non-)presence of breadcrumbs.

So far we were unsuccessful collecting the validation data. CTU-CRAS-Norlab's Spots are still fighting with administrative steps for buying them. Teams CoSTAR and MARBLE who have the real robots agreed to help collecting the validation data, but it is unclear whether they will succeed in doing so anytime soon, probably because of COVID-19 restrictions. In the worst case, we have arranged a rental of Spot for end of April/beginning of May, so this is the latest time when the validation data will be provided for sure. This approach was allowed in an email communication with Angela Maio on 18 Feb 2021:

> model validation data can be submitted up to 3 months after submission of the model (in accordance with the "Model Validation Sessions" contract deliverable)

The design of flippers and robot body have been already finished, while the top deck and sensor box might still undergo minor changes. I'll try to keep the model up-to-date with latest iterations with the manufacturer (until the PR is merged).

We have already uploaded the validation data ZIP with datasheets of our sensor config's sensors. As said earlier, validation data are missing at the moment. It is impossible to find any detailed specs of the Spot's sensors, so their simulation and specification is just an approximation of what could be devised looking at the robot and data. We used the following information sources:

- https://github.com/clearpathrobotics/spot_ros/blob/0cfb0c1c43460238bc2ca113b1d567d6dd1c1985/spot_driver/scripts/spot_wrapper.py#L28 for the list of cameras
- https://github.com/boston-dynamics/spot-sdk/blame/master/python/bosdyn-client/README.md#L58 for a list of resolutions of the onboard cameras
- the Clearpath URDF model
- https://github.com/tu-darmstadt-ros-pkg/image_projection/tree/master/image_projection_demo/doc for sample camera images

The robot can be controlled by any force- or postional- control gait generation algorithm. For ease of use, we prepared https://github.com/ctu-vras/subt_spot package which uses CHAMP to generate a simple gait that works on flat ground (it is open-loop, so terrain will be problematic). As this package depends on unreleased ROS packages (CHAMP), we did not include it directly in the SubT repo. Instead, the users should install this package and its dependencies manually in the solution containers.

The model.sdf files should not be edited manually. The model configuration process of Spot starts with Xacro files in `urdf/` folder and config files in `config/model`. A combination of these files yields an intermediate URDF, which is converted to SDF via `ign sdf -p` and then polished and postprocessed using `scripts/clean_sdf`. This whole conversion process is done in script `scripts/update_robot_sdf`. If a change in BOSDYN_SPOT should affect all sensor configs of Spot, the `update_robot_sdf` script has to be called in each of the packages separately.

There is no URDF file as it is generated on-the-fly by calling `scripts/print_robot_urdf` script. This script is utilized by `launch/description.launch`.

The additional sensor configs reuse most of the code and configuration of the base robot and sensor config 1 instead of copy-pasting it.

---

I was looking into porting gazebo_ros_control to work with Ignition gazebo, but I got quickly completely lost. The main problem is the missing API plugin which would wait for roscore and such things and, most importantly, read and parse the remapping arguments. It seems Ignition Gazebo doesn't yet support plugins that could read the CLI args. That is mostly a show-stopper for any use of ROS in IG. So I ended up writing double-layered plugins - one reads the components of simulator and publishes them on an Ign topic, and the other reads the Ign topic and passes it to ROS. This is not optimal, but as said, there's no way so far to run a ROS publisher inside IG (or I haven't found it).

These plugins might be of more general interest, but due to time constraints, I published them as a part of the `bosdyn_spot` package:

- `configurable_joint_position_controller`
- `logical_contact_system`
- `mono_camera_system`

The `mono_camera_system` works around the non-support of mono cameras in IG. It reads the RGB camera and exposes a new topic `mono/image` which contains the RGB image converted to 8-bit mono. I used the wikipedia-style conversion

    L = 0.3*R + 0.6*G + 0.1*B

But the result is weird in places lit by a spot light. I'm not sure if some gamma magic doesn't need to be done before/after the conversion.

peci1 commented:
This is a preview of all data produced by the base platform, together with a preview of the walking capabilities:

https://user-images.githubusercontent.com/182533/112774097-ca356f80-9038-11eb-90e4-b476884e51db.mp4

peci1 commented:
Mobility of the virtual robot together with an example odometry (in this case only `base_link`->`base_footprint`).


https://user-images.githubusercontent.com/182533/112774185-200a1780-9039-11eb-9e66-f69f1583f8c4.mp4

peci1 commented:
A few Gazebo Classic screenshots that might come handy for model assessment.

Please note that the light visuals are affected by the not yet fixed bug https://github.com/osrf/gazebo/pull/2947 .

![default_gzclient_camera(1)-2021-03-29T02_55_11 806444](https://user-images.githubusercontent.com/182533/112774745-dfab9900-903a-11eb-92d4-bf12611979b7.jpg)
![default_gzclient_camera(1)-2021-03-29T02_55_29 537756](https://user-images.githubusercontent.com/182533/112774747-e0442f80-903a-11eb-9ba2-f970e0248f8e.jpg)
![default_gzclient_camera(1)-2021-03-29T02_55_36 656712](https://user-images.githubusercontent.com/182533/112774749-e0dcc600-903a-11eb-9926-342c861d5a82.jpg)
![default_gzclient_camera(1)-2021-03-29T02_56_52 454448](https://user-images.githubusercontent.com/182533/112774751-e0dcc600-903a-11eb-8898-a3ecdff1d8f2.jpg)
![default_gzclient_camera(1)-2021-03-29T02_57_40 118042](https://user-images.githubusercontent.com/182533/112774752-e1755c80-903a-11eb-94bb-ac5e714b37a9.jpg)
![default_gzclient_camera(1)-2021-03-29T02_58_21 856081](https://user-images.githubusercontent.com/182533/112774753-e1755c80-903a-11eb-9889-cf58bec99915.jpg)

peci1 commented:
Proof of working breadcrumb deploy via ROS topic for sensor config 2 :) 

![2021-03-29T00:31:31 955197026](https://user-images.githubusercontent.com/182533/112774831-344f1400-903b-11eb-883e-41c3b5746afd.png)

peci1 commented:
Preview of sensors available on CTU_CRAS_NORLAB_SPOT_SENSOR_CONFIG_1:

![spot_ctu](https://user-images.githubusercontent.com/182533/112774901-652f4900-903b-11eb-993d-85a59e2d9bda.png)

Please note that "Omni Front" and "Omni Rear" cameras are upside-down on purpose. This is how the real cameras will be mounted.

peci1 commented:
This is the glitch in mono camera rendering I talked about in the first post:

![mono_glitch](https://user-images.githubusercontent.com/182533/112775096-1e8e1e80-903c-11eb-97a7-8e6ca3a3dc43.png)
![rgb_glitch](https://user-images.githubusercontent.com/182533/112775156-441b2800-903c-11eb-82ad-29654cdfc4e7.png)

It seems to me that the black parts are mostly in places with strong spotlight illumination. However, it should not be a problem as the mono cameras are just a postprocessing of RGB images...

peci1 commented:
I downgraded model.sdf to SDF version 1.6 as requested for all our models.

acschang commented:
The initial assessment of the Boston Dynamics' Spot model and associated CTU-CRAS-NORLAB payload is complete however there are some issues to resolve before this model can be merged and utilized in a competitive setting:

Base Spot:
* Please update BOSDYN_SPOT's IMU parameters to match the [new standard SubT IMU parameters](https://github.com/osrf/subt/pull/798).
* Please add intrinsic values for all camera sensors.
* Please confirm the robot mass and make sure the mass in `specifications.md` and `model.sdf` agree.
* Please update the endurance to meet the value specified in the `specifications.md`.  This can be done by adjusting the `power_load` parameter in `common.rb` to 6.6 as opposed to 4.95.
* Please provide validation data as it is necessary to complete the evaluation. If not provided, the model's maximum velocity will be limited to 1m/s and battery life set to 60min. When the model validation data can be provided later, please submit a PR updating the model.

CTU-CRAS-NORLAB Payload:
* Please add intrinsic values for all cameras sensors in the OmniCamera.
* Please update the Ouster range resolution and stddev values to match the standard in the SubT repository (0.01 for both).
* Please update IMU parameters on the CTU-CRAS-NORLAB sensor payload to match the [new standard SubT IMU parameters](https://github.com/osrf/subt/pull/798).

Please address the issues noted above in this pull request. Once the above issues are fixed, your submission will move onto the next review stage!

mjcarroll commented on submitted_models/bosdyn_spot/specifications.md:
```diff
@@ -0,0 +1,235 @@
+The simulator publishes information about contact of the lower legs with terrain. There are two
+types of topics. Both depend on the presence of plugin `ignition-gazebo-contact-system` in the
+simulation world. This has been added to the simulation worlds in https://github.com/osrf/subt/pull/834 .
+The PR has not yet been merged. For preliminary testing, you can try running `bosdyn_spot/launch/example.ign`
```
Nit: It's merged now.

peci1 replied:
Reworded.

mjcarroll commented on submitted_models/bosdyn_spot/specifications.md:
```diff
@@ -0,0 +1,235 @@
+
+To control each joint separately via commanded effort, there are topics `cmd_force/front_left_hip_x` etc.
+of type `std_msgs/Float64`. There might be time sync problems when commanding each joint separately, so
+it is suggested to not used this interface and use the multi-commanding topic described further.
```
This paragraph is duplicated in the next paragraph.

peci1 replied:
It's not an exact duplicate. The first one talks about effort-control and the second one about position control. I'll try to reword it so that it looks less confusing.

mjcarroll commented on subt_ros/src/SubtRosRelay.cc:
```diff
@@ -234,7 +234,7 @@ SubtRosRelay::SubtRosRelay()
   recorderOptions.append_date=false;
   recorderOptions.split=true;
-  recorderOptions.max_splits=1;
+  recorderOptions.max_splits=0;
```
Let's remove this change.

peci1 replied:
Outch, this should have stayed local :)

mjcarroll reviewed (commented):
Quick pass while waiting for it to build.

I didn't really inspect your system plugins for correct behavior, but they don't use any ground truth information, so they should be suitable for competition.

peci1 commented:
I addressed all of the issues except validation data, which should however be already being collected by CoSTAR.

Base Spot:

 - [x] Please update BOSDYN_SPOT's IMU parameters to match the new standard SubT IMU parameters.
 - [x] Please add intrinsic values for all camera sensors.
 - [x] Please confirm the robot mass and make sure the mass in specifications.md and model.sdf agree.
   - the model I based the work on apparently did not model the 4 kg battery. I added it. That required tuning of the PIDs for CTU_CRAS_NORLAB_SPOT as with the 4 kg more it became very unstable.
 - [x] Please update the endurance to meet the value specified in the specifications.md. This can be done by adjusting the power_load parameter in common.rb to 6.6 as opposed to 4.95.
 - [ ] Please provide validation data as it is necessary to complete the evaluation. If not provided, the model's maximum velocity will be limited to 1m/s and battery life set to 60min. When the model validation data can be provided later, please submit a PR updating the model.

CTU-CRAS-NORLAB Payload:

 - [x] Please add intrinsic values for all cameras sensors in the OmniCamera.
 - [x] Please update the Ouster range resolution and stddev values to match the standard in the SubT repository (0.01 for both).
 - [x] Please update IMU parameters on the CTU-CRAS-NORLAB sensor payload to match the new standard SubT IMU parameters.


The body mono camera intrinsics are based on this camera info from the real Spot provided by CoSTAR:

```

header: 
  seq: 214
  stamp: 
    secs: 1603401456
    nsecs: 366378422
  frame_id: "back_fisheye"
height: 480
width: 640
distortion_model: "plumb_bob"
D: [0.0, 0.0, 0.0, 0.0, 0.0]
K: [257.48651123046875, 0.0, 316.9465637207031, 0.0, 256.9732971191406, 245.06680297851562, 0.0, 0.0, 1.0]
R: [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]
P: [257.48651123046875, 0.0, 316.9465637207031, 0.0, 0.0, 256.9732971191406, 245.06680297851562, 0.0, 0.0, 0.0, 1.0, 0.0]
binning_x: 0
binning_y: 0
roi: 
  x_offset: 0
  y_offset: 0
  height: 0
  width: 0
  do_rectify: False
```

I chose the use the ideal values for `cx` and `cy` though, as it doesn't really make sense to use other values in the simulation.

The depth camera intrinsics are so far based on the mono camera, just different resolution. When I get my hands on a camera info from these, I'll update the models to be as close to the real ones as possible.

I also added a link to experimental validation data for Basler cameras on CTU_CRAS_NORLAB_SPOT.

acschang commented on submitted_models/bosdyn_spot/launch/common.rb:
```diff
@@ -91,7 +91,7 @@ def _spawner(_name, _modelURI, _worldName, _x, _y, _z, _roll, _pitch, _yaw, _add
         <resistance>0.061523</resistance>
         <smooth_current_tau>1.9499</smooth_current_tau>
-        <power_load>4.95</power_load>
+        <power_load>6.6</power_load>
```
The `power_load` should be 9.9.  My apologies for giving you the incorrect value.  Your adjustment on your other PR caught this.

peci1 replied:
Fixed.

acschang commented on submitted_models/bosdyn_spot/urdf/camera.xacro:
```diff
@@ -20,7 +20,7 @@
                     <visualize>${visualize}</visualize>
                     <camera>
-                        <horizontal_fov>${radians(95)}</horizontal_fov>
+                        <horizontal_fov>${radians(102.35)}</horizontal_fov> <!-- computed from real fx -->
```
Is the update to the camera FoV based upon information from CoSTAR or something else?

peci1 replied:
Yes, it's computed from the `fx` provided by CoSTAR.

mjcarroll commented on submitted_models/ctu_cras_norlab_spot_sensor_config_2/launch/spawner.rb:
```diff
@@ -0,0 +1,44 @@
+  # description.launch is included from vehicle_topics.launch
+  <<-HEREDOC
+  <executable name='topics'>
+    <command>roslaunch --wait ctu_cras_norlab_spot_sensor_config_1 vehicle_topics.launch world_name:=#{_worldName} name:=#{_name} breadcrumbs:=#{$num_breadcrumbs} description_print_command:="#{File.dirname(__FILE__) + '/../scripts/print_robot_urdf'}"</command>
```
Should this be config_2?

acschang replied:
No, this is correct.  The launch files are nested and the breadcrumb parameter percolates down to the [breadcrumbs launch file](https://github.com/osrf/subt/blob/master/subt_ros/launch/models_common/breadcrumbs.launch) called from `BOSDYN_SPOT`'s `vehicle_topics.launch`.  I experimentally verified that breadcrumbs function for this model and do not exist on sensor config 1.

mjcarroll commented on submitted_models/ctu_cras_norlab_spot_sensor_config_2/model.config:
```diff
@@ -0,0 +1,14 @@
+  </description>
+  <depend>
+    <model>
+      <uri>https://fuel.ignitionrobotics.org/1.0/OpenRobotics/models/CTU_CRAS_NORLAB_SPOT_SENSOR_CONFIG_1</uri>
```
SENSOR_CONFIG_2

acschang replied:
We don't really have a precedent for this case as normally sensor config 2 does not have its own `model.sdf` file however the _package_ certainly depends on sensor config 1 being present as its `vehicle_topics.launch` is leveraged.

Another reason there doesn't appear to be a precedent for this is that `model.config` files seem to be exceedingly rare in breadcrumb sensor configs.

peci1 replied:
There is a precedent now :) https://github.com/osrf/subt/blob/master/submitted_models/ctu_cras_norlab_marv_sensor_config_2/model.config

mjcarroll commented on submitted_models/ctu_cras_norlab_spot_sensor_config_2/model.sdf:
```diff
@@ -0,0 +1,2665 @@
+      <collision name="base_link_collision">
+        <geometry>
+          <mesh>
+            <uri>../bosdyn_spot/meshes/body_collision.stl</uri>
```
These relative paths are going to break when the models are uploaded to fuel, because of the additional version number directory.

I would propose uploading the current SPOT configuration, and adjusting these URIs to be correct for Fuel-based models.  After this PR is approved and lands, we can reupload the SPOT model with the final approved setup.

What do you think @peci1 ?

CC: @azeey, who is also reviewing a SPOT-derived model.

mjcarroll replied:
I went ahead an uploaded as: https://app.ignitionrobotics.org/OpenRobotics/fuel/models/bosdyn_spot

peci1 replied:
Thanks a lot, I got a bit overwhelmed by other tasks yesterday... I'll have a look at your other comments in a few hours

mjcarroll replied:
Alright, I think this is the only thing remaining from my pov.

peci1 replied:
I'll go ahead and do the URL unfurling also for references to CTU_CRAS_NORLAB_SPOT_SC1, which will be shortly invalid (until the model is uploaded to Fuel).

peci1 replied:
I tested sensor config 1 with the new change and it loaded correctly.

peci1 commented:
Sensor config 2 also works if I manually add the SC1 files in `.ignition/fuel` cache. What is surprising (or not at all?) is that the previous relative references worked too, in this case (i.e. Fuel had to add the `/1` version directory to the mesh URL itself). So maybe the relative references would work, too, but now when the models will already be uploaded to Fuel, having the absolute URLs is better.

peci1 commented:
When uploading Sensor Config 1 to Fuel, don't forget that most DAE files in the meshes directory are actually symlinks to bosdyn_spot meshes. So be sure to follow symlinks when uploading.

mjcarroll commented:
@acschang anything else from you?

mjcarroll commented:
@peci1 as a follow up, please verify: https://app.ignitionrobotics.org/OpenRobotics/fuel/models/CTU_CRAS_NORLAB_SPOT_SENSOR_CONFIG_1

and

https://app.ignitionrobotics.org/OpenRobotics/fuel/models/CTU_CRAS_NORLAB_SPOT_SENSOR_CONFIG_2

nkoenig commented:
Bounding box size: `[0.87244, 0.448977, 1.07188]`

peci1 commented:
@mjcarroll I tested both sensor configs and they work well. I tried  clearing them from local Fuel cache and starting with sensor config 2, and it correctly downloaded both sensor config 1 and bosdyn_spot.

I've discovered a problem with the use of nodelets, though. See #898 where I disabled running nodelets in nodelet managers to get a more reliable behavior.

peci1 commented:
I've seen you assigned CTU-CRAS-Norlab Spot a credit price of 300 credits (on https://app.ignitionrobotics.org/OpenRobotics/fuel/models/CTU_CRAS_NORLAB_SPOT_SENSOR_CONFIG_1 ). I think it is a lot compared to other robots and taking into account how difficult it is to control this robot and how easy it is to tip it over (I haven't found a way to stand it once it falls). If possible, please, reconsider the price, otherwise Spot is practically useless...

peci1 commented:
Thanks, I see Angela decreased the cost to 255/280!
